### PR TITLE
Pod Scheduling Readiness

### PR DESCRIPTION
**Pull request by ykcai-daniel**
Updated PR of #3612, a gated task no longer blocks the job. Do not use new state.

### CODE REVIEW

File: pkg/scheduler/api/job_info.go [hunk 1]
```diff
@@ -575,6 +590,17 @@ func (ji *JobInfo) DeleteTaskInfo(ti *TaskInfo) error {
 	return nil
 }
 
+func (ji* JobInfo) HaveSchGatedTask() bool{
+	for _,task:=range ji.Tasks{
```
Comment (googs1025) [inline pkg/scheduler/api/job_info.go:594]: nit: code style, need space

**Comment by googs1025** [comment]
Does this change have the same purpose as this PR https://github.com/volcano-sh/volcano/pull/3612
?

**Comment by Monokaix** [comment]
Hi，I think there is no necessary to submit two PRs, just based the previous PR and update it OK，and please also update the design doc.

**Comment by Monokaix** [comment]
Please refer to kube-scheduler to add a featuregate to enable/disable this feature.

File: pkg/scheduler/api/job_info.go (hunk 1 above)
Comment (ykcai-daniel) [reply pkg/scheduler/api/job_info.go]: Fixed

**Comment by ykcai-daniel** [comment]
@Monokaix PR updated. I will also update the design proposal #3581 to document the changes.

**Comment by ykcai-daniel** [comment]
PR summary: Pod Scheduling gates is a new feature in K8S, which uses the `.spec.schedulingGates` field of Pod to signal that a Pod is not ready for scheduling (Pod will be in pending state when gated). This feature is often used to implement customized resource managers and we want to support it for compatibility. Main Changes involve:

1.  Scheduling Actions: Job with scheduling gated tasks mainly exist in the Inqueue state. In allocate action, if the pod of a task is scheduling gates, it will be skipped for allocation and not bound to a node. Consequently, if too many tasks are scheduling gates, the job will be gang-unschedulable
2. Plugins: Since scheduling gated pods are not ready to be allocated, we don't scheduling gated tasks to consume inqueue resources, making other jobs uninqueuable. Therefore, proportion, capacity and overcommit plugins have to be changed to deduct the resources of scheduling gated tasks
3. Pod Conditions: When each session close, the Condition of scheduling gated pods will not be updated, showing the Schedulable False condition with reason ScheduleGated, which is created by api-server

Future Works that are not covered in  this PR:

1. Now, we do not support removing scheduling gates by modifying the template of a vcjob because supporting this feature involve major changes to controller and Pod-level gates removal is sufficient for most usecase. In the future, if there is a need for task template support, we can add it.
2. Events and Correct Message: Now, if a Job is gang unschedulable due to scheduling gates, the reason in condition is "NoEnoughResources". We might need better reason and message for scheduling gates.

**Comment by ykcai-daniel** [comment]
Reviewer: /assign @Monokaix /assign @googs1025

File: pkg/scheduler/plugins/capacity/capacity.go [hunk 2]
```diff
@@ -142,7 +142,7 @@ func (cp *capacityPlugin) OnSessionOpen(ssn *framework.Session) {
 		}
 
 		if job.PodGroup.Status.Phase == scheduling.PodGroupInqueue {
-			attr.inqueue.Add(job.GetMinResources())
+			attr.inqueue.Add(job.DeductSchGatedResources(job.GetMinResources()))
```
Comment (lowang-bh) [inline pkg/scheduler/plugins/capacity/capacity.go:145]: I think its better to do the calcalation in GetMinResources function. Let GetMinResources return the min resource to run the job.

File: pkg/scheduler/api/job_info.go [hunk 3]
```diff
@@ -228,6 +230,17 @@ func (ti *TaskInfo) ClearLastTxContext() {
 	ti.LastTransaction = nil
 }
 
+// Return if the pod of a task is scheduling gated by checking if length of sch gates
+// When the Pod is not yet created or sch gates field not set, return false
+func (ti *TaskInfo) SchedulingGated() bool {
```
[the remark this answers is not in the document]
Comment (ykcai-daniel) [reply pkg/scheduler/api/job_info.go]: I have added a field `schGated` into taskInfo

**Comment by ykcai-daniel** [comment]
@lowang-bh cc @Monokaix  Pull request updated. Please review the changes. Thank you!

File: pkg/scheduler/api/job_info.go [hunk 4]
```diff
@@ -228,6 +233,22 @@ func (ti *TaskInfo) ClearLastTxContext() {
...
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodSchedulingReadiness) {
+		return pod != nil && pod.Spec.SchedulingGates != nil && len(pod.Spec.SchedulingGates) != 0
+	}
+	return false
+
+}
+
+// ti.SchGated is initialized in NewTaskInfo
+func (ti *TaskInfo) SchedulingGated() bool {
+	return ti.SchGated
```
Comment (Monokaix) [inline pkg/scheduler/api/job_info.go:249]: Should this filed be exported?

File: pkg/scheduler/plugins/capacity/capacity.go [hunk 5]
```diff
@@ -282,18 +282,18 @@ func (cp *capacityPlugin) OnSessionOpen(ssn *framework.Session) {
 		queue := ssn.Queues[queueID]
 		// If no capability is set, always enqueue the job.
 		if attr.realCapability == nil {
-			klog.V(4).Infof("Capability of queue <%s> was not set, allow job <%s/%s> to Inqueue.",
+			klog.V(3).Infof("Capability of queue <%s> was not set, allow job <%s/%s> to Inqueue.",
```
Comment (Monokaix) [inline pkg/scheduler/plugins/capacity/capacity.go:285]: Why change these log level?

Comment (ykcai-daniel) [reply pkg/scheduler/plugins/capacity/capacity.go:285]: I changed this log level when debugging the capcacity plugin. Will change is back

File: pkg/scheduler/api/job_info.go (hunk 4 above)
Comment (ykcai-daniel) [reply pkg/scheduler/api/job_info.go:249]: With this function, there is no need to export it. I will change it to `schGated`.

File: pkg/scheduler/api/job_info.go [hunk 6]
```diff
@@ -504,6 +526,39 @@ func (ji *JobInfo) GetMinResources() *Resource {
...
+		}
+	}
+	return res
+}
+
+// Deduct resources of scheduling gated pod from Resource res;
+// If resource is less than gated resources, return zero;
+// Gated resources might be more than minResource when total rep of each task > minAvailable
+func (ji *JobInfo) DeductSchGatedResources(res *Resource) *Resource {
+	result := res.Clone()
```
Comment (Monokaix) [inline pkg/scheduler/api/job_info.go:545]: We can check whether res Is zero to have a shortcut here, because in most case task is not gated.

File: pkg/scheduler/api/job_info.go (hunk 4 above)
Comment (lowang-bh) [reply pkg/scheduler/api/job_info.go:249]: I would like to use this filed instead of a function calling.

Comment (Monokaix) [reply pkg/scheduler/api/job_info.go:249]: It's an optional way, but seems there is no need to expose an extra filed of task.

**Comment by ykcai-daniel** [comment]
For more details about the design of the PR, see #3581 . Please also merge the design doc.

File: pkg/scheduler/api/job_info.go (hunk 4 above)
Comment (Monokaix) [reply pkg/scheduler/api/job_info.go]: I think it's reasonable.

File: pkg/scheduler/api/job_info.go [hunk 7]
```diff
@@ -504,6 +526,44 @@ func (ji *JobInfo) GetMinResources() *Resource {
...
+		}
+	}
+	return res
+}
+
+// Deduct resources of scheduling gated pod from Resource res;
+// If resource is less than gated resources, return zero;
+// Note: The purpose of this functionis to deduct the resources of scheduling gated tasks
+// in a job when calculating inqueued resources so that it will not block other jobs from being inqueued.
+func (ji *JobInfo) DeductSchGatedResources(res *Resource) *Resource {
```
Comment (Monokaix) [inline pkg/scheduler/api/job_info.go:545]: The comment should start with `DeductSchGatedResources` to be consistent with func name.

File: pkg/scheduler/api/job_info.go [hunk 8]
```diff
@@ -504,6 +526,44 @@ func (ji *JobInfo) GetMinResources() *Resource {
...
+	result := res.Clone()
```
Comment (Monokaix) [inline pkg/scheduler/api/job_info.go:546]: Seems there is no need to clone `res` because no one will also change it, and also we can change to:

```go
schGatedResource := ji.GetSchGatedPodResources()
// Most jobs do not have any scheduling gated tasks, hence we add this short cut
if schGatedResource.IsEmpty() {
    return res
}
```

File: pkg/scheduler/api/job_info.go [hunk 9]
```diff
@@ -576,6 +636,15 @@ func (ji *JobInfo) DeleteTaskInfo(ti *TaskInfo) error {
 	return nil
 }
 
+func (ji *JobInfo) HaveSchGatedTask() bool {
```
Comment (Monokaix) [inline pkg/scheduler/api/job_info.go:639]: Is this called somewhere?

File: pkg/scheduler/api/resource_info_test.go [hunk 10]
```diff
@@ -818,21 +818,6 @@ func TestLessEqualWithDimension(t *testing.T) {
 			},
 			expected: true,
 		},
-		{
```
Comment (Monokaix) [inline pkg/scheduler/api/resource_info_test.go:821]: Why is this removed?

Comment (ykcai-daniel) [reply pkg/scheduler/api/resource_info_test.go]: Change Reverted

File: pkg/scheduler/api/job_info.go (hunk 9 above)
Comment (ykcai-daniel) [reply pkg/scheduler/api/job_info.go]: Redundent function removed.

File: pkg/scheduler/api/job_info.go (hunk 8 above)
Comment (ykcai-daniel) [reply pkg/scheduler/api/job_info.go]: Now, we don't clone `res` if scheduling gated resource is zero.

File: test/e2e/util/deployment.go [hunk 11]
```diff
@@ -29,8 +29,12 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 )
 
-// CreateDeployment creates a new deployment
 func CreateDeployment(ctx *TestContext, name string, rep int32, img string, req v1.ResourceList) *appv1.Deployment {
+	return CreateDeploymentGated(ctx, name, rep, img, req, []v1.PodSchedulingGate{})
+}
+
+// CreateDeployment creates a new deployment
```
Comment (Monokaix) [inline test/e2e/util/deployment.go:36]: CreateDeployment -》 CreateDeploymentGated
